metrics: Add the number of requests in the queue

vadmeste opened the pull request:
## Description
We can use this metric to check if there are too many S3 clients in the
queue and could explain why some of those S3 clients are timing out.

```
minio_s3_requests_waiting_total{server="127.0.0.1:9000"} 9981
```

If max_requests is 10000 then there is a strong possiblity that clients
are timing out because of the queue deadline

## Motivation and Context
Add new metric showing the number of clients in the pipeline

## How to test this PR?
```
$ MINIO_PROMETHEUS_AUTH_TYPE="public" minio server /tmp/xl/{1...4}/
$ mc mb myminio/testbucket/
$ mc admin config set myminio/ api max_requests=10
$ mc cp -r /usr/share/ myminio/testbucket/
$ curl http://localhost:9000/minio/v2/metrics/cluster | grep wait 
```
## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -259,9 +259,10 @@ type ServerHTTPAPIStats struct {
-	CurrentS3Requests ServerHTTPAPIStats `json:"currentS3Requests"`
-	TotalS3Requests   ServerHTTPAPIStats `json:"totalS3Requests"`
-	TotalS3Errors     ServerHTTPAPIStats `json:"totalS3Errors"`
+	S3RequestsInPipeline int32              `json:"s3RequestsInPipeline"`
```
```suggestion
	S3RequestsInQueue int32              `json:"s3RequestsInPipeline"`
```

klauspost commented on cmd/handler-api.go:
```diff
@@ -141,7 +141,11 @@ func maxClients(f http.HandlerFunc) http.HandlerFunc {
+				<-pool
+				globalHTTPStats.addRequestsInQueue(-1)
+			}()
+			globalHTTPStats.addRequestsInQueue(1)
```
This is incorrect. You are counting running requests, not queued requests.

Queued requests are blocked on `pool <- struct{}{}` so they will only arrive here when executing.

klauspost replied:
Could just be semantics, but `s3RequestsInQueue` should be `s3RequestsRunning` or similar.

vadmeste replied:
You are right, it is fixed now, but still keeping queue semantic, please review again :)

klauspost commented on cmd/handler-api.go:
```diff
@@ -136,20 +136,25 @@ func maxClients(f http.HandlerFunc) http.HandlerFunc {
 		select {
 		case pool <- struct{}{}:
 			defer func() { <-pool }()
+			globalHTTPStats.addRequestsInQueue(-1)
```
Defer these when the queue count is incremented.

That way even a panic will get the count properly decremented.

vadmeste replied:
> Defer these when the queue count is incremented.

It will be wrong, we need to decrement when a request is out from the pool, otherwise we won't have the exact numbers of clients in the pool.

> That way even a panic will get the count properly decremented.

panic will happen in the next line, ServerHTTP() and we already decremented at this stage.

klauspost replied:
Make sense. Let's keep it!

klauspost reviewed (approved):
LGTM

Commit:
metrics: Add the number of requests in the waiting queue

We can use this metric to check if there are too many S3 clients in the
queue and could explain why some of those S3 clients are timing out.

```
minio_s3_requests_waiting_total{server="127.0.0.1:9000"} 9981
```

If max_requests is 10000 then there is a strong possiblity that clients
are timing out because of the queue deadline.
